lib/teleterm/services/desktop/directorysharing.go

```diff
@@ -0,0 +1,296 @@
+	// Switch to it when branch/v17 is updated to Go 1.24.
+	safePathGetter func(relativePath string) (string, error)
+}
+
+// FileOrDirInfo contains metadata about a file or a directory.
+type FileOrDirInfo struct {
+	Size         int64
+	LastModified int64
+	FileType     FileType // "file" or "directory"
+	IsEmpty      bool
+	Path         string
+}
+
+type FileType uint32
+
+const (
+	FileTypeFile FileType = iota
+	FileTypeDir
+)
+
+const StandardDirSize = 4096
+
+func (s *DirectoryAccess) Open(baseDir string) error {
+	if s.safePathGetter != nil {
+		return errors.New("only one shared directory is supported")
+	}
+
+	base, err := filepath.EvalSymlinks(baseDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(base)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !stat.IsDir() {
+		return trace.BadParameter("%q is not a directory", baseDir)
+	}
+
+	s.safePathGetter = func(relativePath string) (string, error) {
+		full := filepath.Join(base, relativePath)
+		resolved, err := filepath.EvalSymlinks(full)
+		if err != nil {
+			// EvalSymlinks returns an error if the target file does not exist.
+			// In that case, attempt to resolve the symlinks of the parent directory instead.
+			if os.IsNotExist(err) {
+				parent := filepath.Dir(full)
+				resolvedParent, perr := filepath.EvalSymlinks(parent)
+				if perr != nil {
+					return "", trace.Wrap(perr)
+				}
+
+				// Reconstruct the full path by joining the resolved parent with the original file name.
+				resolved = filepath.Join(resolvedParent, filepath.Base(full))
+			} else {
+				return "", trace.Wrap(err)
+			}
+		}
+		if !isSubPath(base, resolved) {
+			return "", trace.BadParameter("path escapes from parent")
+		}
+		return resolved, nil
+	}
+	return nil
+}
+
+func isSubPath(parent, child string) bool {
+	return child == parent || strings.HasPrefix(child, parent+string(filepath.Separator))
+}
+
+// Stat retrieves metadata about a file or directory at the given path.
+func (s *DirectoryAccess) Stat(relativePath string) (*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := s.readFileOrDirInfo(relativePath, stat)
+	return info, trace.Wrap(err)
+}
+
+// ReadDir lists files and directories within the given directory path, skips symlinks.
+func (s *DirectoryAccess) ReadDir(relativePath string) ([]*FileOrDirInfo, error) {
+	path, err := s.safePathGetter(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var results []*FileOrDirInfo
+	for _, entry := range entries {
+		fileInfo, err := entry.Info()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Skip symlinks, we can't present them properly in the remote machine.
```

Comment:
Do we skip symlinks in the web UI too?

Reply:
@zmb3 I don't see them in Chrome. They don’t appear to be supported by the File System Access API https://github.com/WICG/file-system-access/issues/113#issuecomment-2379658548.

Actually, now that I think about it, we could resolve the symlinks if they are "safe" and just return the resolved file or directory.
The downside is that users won’t be able to tell whether the item was originally a symlink or a regular file/directory, which could be confusing. What do you think?

Reply:
After discussing it on 1x1: we will not show symlinks to keep the same behavior as in Web UI.
We can change it if needed in the future.